File: chatbot/src/chatbot.js
import neuro from 'neuro.js';
import natural from 'natural';
import nlp from 'compromise';


var TextClassifier = neuro.classifiers.multilabel.BinaryRelevance.bind(0, {
    binaryClassifierType: neuro.classifiers.Winnow.bind(0, { retrain_count: 10 })
});


var WordExtractor = function (input, features) {
    const tokenizer = new natural.WordTokenizer();
    tokenizer.tokenize(input.toLowerCase()).forEach(function (word) {
        features[word] = 1;
    });
};


var intentClassifier = new neuro.classifiers.EnhancedClassifier({
    classifierType: TextClassifier,
    featureExtractor: WordExtractor
});



function train() {
    console.log("train")
    intentClassifier.trainBatch([
        { input: "Como está o tempo hoje?", output: ["current_weather", "temperature_forecast"] },
        { input: "Qual é a previsão do tempo para amanhã?", output: ["weather_forecast"] },
        { input: "Está chovendo agora?", output: ["current_weather"] },
        { input: "Vai chover amanhã?", output: ["weather_forecast"] },
        { input: "Qual a temperatura atual?", output: ["current_temperature"] },
        { input: "Qual será a temperatura amanhã?", output: ["temperature_forecast"] },
        { input: "Faz calor amanhã?", output: ["temperature_forecast"] },
        { input: "Faz frio aqui", output: ["current_weather", "current_temperature"] },
        { input: "Vai nevar em São Paulo na quarta?", output: ["weather_forecast"] },
        { input: "Qual é a temperatura agora em Nova York?", output: ["current_temperature"] },
        { input: "Vai chover no Rio de Janeiro amanhã?", output: ["weather_forecast"] },
        { input: "Está nublado em São Paulo?", output: ["current_weather"] },
        { input: "Vai fazer frio na quarta-feira?", output: ["temperature_forecast"] },
        { input: "Previsão do tempo para hoje à noite", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Quantos graus está fazendo agora?", output: ["current_temperature"] },
        { input: "Qual será a máxima de amanhã?", output: ["temperature_forecast"] },
        { input: "Quais são as chances de chuva hoje?", output: ["current_weather", "weather_forecast"] },
        { input: "Como estará o tempo no fim de semana?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Vai nevar no inverno?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Qual é a sensação térmica agora?", output: ["current_temperature"] },
        { input: "O que esperar do clima para a próxima semana?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Como estará o clima no Natal?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Vai ventar forte hoje?", output: ["current_weather"] },
        { input: "A umidade está alta agora?", output: ["current_weather"] },
        { input: "Quando será a próxima tempestade?", output: ["weather_forecast"] },
        { input: "Qual será a temperatura na semana que vem?", output: ["temperature_forecast"] },
        { input: "Vai chover na semana que vem?", output: ["weather_forecast"] },
        { input: "Como estava o tempo no mês passado?", output: ["current_weather"] },
        { input: "Qual era a temperatura no mês passado?", output: ["current_temperature"] },
        { input: "Qual é a previsão do tempo para a próxima semana?", output: ["weather_forecast"] },
        { input: "Vai nevar na próxima semana?", output: ["weather_forecast"] },
        { input: "Como estará o clima na próxima semana?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Estava muito quente no mês passado?", output: ["current_temperature"] },
        { input: "Qual será a máxima da próxima semana?", output: ["temperature_forecast"] },
        { input: "A previsão para a semana que vem é de chuva?", output: ["weather_forecast"] },
        { input: "A previsão para a próxima semana é de calor intenso?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Semana que vem vai fazer tanto calor quanto fez hoje?", output: ["weather_forecast", "temperature_forecast", "current_temperature"] },
        { input: "Amanhã qual será a temperatura em São Paulo?", output: ["temperature_forecast"] },
        { input: "Amanhã qual a temperatura na Baixada?", output: ["temperature_forecast"] },
        { input: "Amanhã qual a temperatura na Cidade?", output: ["temperature_forecast"] },
        { input: "Semana que vem qual a temperatura na Cidade?", output: ["temperature_forecast"] },
        { input: "Terça feira qual a temperatura no Galeão?", output: ["temperature_forecast"] },
        { input: "Segunda-feira qual a temperatura em Santos Lagos?", output: ["temperature_forecast"] },
        { input: "Na próxima segunda-feira vai chover no Rio de Janeiro?", output: ["weather_forecast"] },
        { input: "No próximo mês teremos dias quentes em Brasília?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Na semana que vem, como estará o clima em Salvador?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Qual é a previsão do tempo para o próximo fim de semana?", output: ["weather_forecast"] },
        { input: "Ontem, como estava o clima em Recife?", output: ["current_weather"] },
        { input: "Próxima quarta-feira terá temperaturas baixas em Florianópolis?", output: ["temperature_forecast"] },
        { input: "No próximo feriado, vai chover em Belo Horizonte?", output: ["weather_forecast"] },
        { input: "No próximo dia útil, qual será a temperatura em Manaus?", output: ["temperature_forecast"] },
        { input: "Daqui a dois dias, como estará o tempo em Porto Alegre?", output: ["weather_forecast", "temperature_forecast"] },
        { input: "Chove hoje em São Paulo?", output: ["weather_forecast"] },
        { input: "Choverá hoje em Juariri?", output: ["weather_forecast"] },
        { input: "Vai chover hoje no Espírito Santo?", output: ["weather_forecast"] },
        { input: "Vai fazer frio sexta feira no Amazonas?", output: ["temperature_forecast"] },
        { input: "Vai fazer calor sábado na Lagoa?", output: ["temperature_forecast"] },
    ]);
}


function extractKeywordsWithLib(question) {
    let doc = nlp(question.toLowerCase());
    console.log('doc');
    console.log(doc.topics().json())
    let dateMatch = doc.dates().out('text');
    let locationMatch = doc.places().out('text');
    let weatherTerms = ["clima", "tempo", "temperatura", "calor", "frio", "chuva", "chove", "chover", "sol", "nublado", "céu", "brisa", "neve", "neva", "nevar", "vento", "ventar", "umidade", "precipitação", "ventania", "graus", "pressão atmosférica", "tempestade", "nascer do sol", "pôr do sol"];
    const hasWeatherTerm = weatherTerms.some(term => doc.has(term));

    return {
        date: dateMatch || null,
        location: locationMatch || null,
        hasWeatherTerm: hasWeatherTerm
    };
}


/**
 * Extracts keywords from a question.
 * @param {string} question - The input question.
 * @returns {Keywords} The extracted keywords.
 */
function extractKeywords(question) {
    const weatherTerms = ["Clima", "Tempo", "Temperatura", "Calor",
        "Frio", "Chuva", "chover", "Chove", "Sol", "Nublado", "Céu", "Brisa",
        "Neve", "nevar", "neva", "Vento", "ventar", "venta", "Umidade", "Precipitação", "Ventania",
        "Graus", "Pressão atmosférica", "Tempestade",
        "nascer do sol", "pôr do sol",];
    const weatherTermsRegex = new RegExp(weatherTerms.join("|"), "gi");

    const dateRegex = /(hoje|manh[aã]|amanh[ãa]|ontem|tarde|cedo|segunda|ter[çc]a|quarta|quinta|sexta|semana que vem|m[eê]s que vem|ano que vem)/;
    let dateMatch = question.toLowerCase().match(dateRegex);

    let locationMatch = question.toLowerCase().replace(weatherTermsRegex, "");

    console.log("locationMatch", locationMatch);
    console.log("dateMatch", dateMatch);
    const beforeRegex = /^.*\b(em|no|vai)\b\s+/i;
    const afterRegex = /\s+\b(na|no|est[áa]|vai|faz)\b.*$/i;
    let tempResult = locationMatch.replace(beforeRegex, '');
    let finalResult = tempResult.replace(afterRegex, '');
    finalResult = finalResult.replace(dateRegex, "");

    const loc = finalResult.trim().replace("?", "");


    const hasWeatherTerm = weatherTerms.some(term => question.toLowerCase().includes(term.toLowerCase()));
    console.log(dateMatch ? dateMatch[0].toLowerCase() : "nao tem date match");
    console.log(loc ? loc : "nao tem location match");
    return {
        date: dateMatch ? dateMatch[0].toLowerCase() : null,
        location: loc ? loc : null,
        hasWeatherTerm: hasWeatherTerm
    };
}

// Generate a random number between min (inclusive) and max (exclusive)
function getRandomNumber(min, max) {
    return Math.floor(Math.random() * (max - min)) + min;
}

/**
 * Generates a response based on the classified intents and extracted keywords.
 * @param {string[]} intents - The classified intents.
 * @param {Keywords} keywords - The extracted keywords.
 * @returns {string} The generated response.
 */
function getWeatherResponse(intents, keywords) {
    const responses = [];
    const local = keywords.location ? ` em ${keywords.location}` : "";
    const date = keywords.date ? `${keywords.date}` : "amanhã";

    const weatherResponses = [
        `O tempo${local} está claro e ensolarado.`,
        `O tempo${local} está nublado com possibilidade de chuvas.`,
        `O tempo${local} está chuvoso e ventoso.`,
        `O tempo${local} está frio e nublado.`
    ];

    const forecastResponses = [
        `A previsão para ${date}${local} é de chuva com possibilidade de trovoadas.`,
        `A previsão para ${date}${local} é de céu claro e ensolarado.`,
        `A previsão para ${date}${local} é de tempo nublado com pancadas de chuva.`,
        `A previsão para ${date}${local} é de tempo frio e ventoso.`
    ];

    const currentTemperatureResponses = [
        `A temperatura atual${local} é de ${getRandomNumber(23, 37)}°C.`,
        `Atualmente, a temperatura${local} está em torno de ${getRandomNumber(12, 37)}°C.`,
        `Agora, a temperatura${local} é de ${getRandomNumber(15, 30)}°C.`,
        `Neste momento, a temperatura${local} está em ${getRandomNumber(18, 33)}°C.`
    ];

    const temperatureForecastResponses = [
        `A temperatura ${date}${local} será em torno de ${getRandomNumber(12, 37)}°C.`,
        `Esperamos uma temperatura de ${getRandomNumber(15, 35)}°C ${date}${local}.`,
        `A previsão de temperatura ${date}${local} é de ${getRandomNumber(10, 25)}°C.`,
        `Para ${date}${local}, a temperatura deve ficar em torno de ${getRandomNumber(20, 30)}°C.`
    ];

    intents.forEach(intent => {
        switch (intent) {
            case 'current_weather':
                responses.push(weatherResponses[Math.floor(Math.random() * weatherResponses.length)]);
                break;
            case 'weather_forecast':
                responses.push(forecastResponses[Math.floor(Math.random() * forecastResponses.length)]);
                break;
            case 'current_temperature':
                responses.push(currentTemperatureResponses[Math.floor(Math.random() * currentTemperatureResponses.length)]);
                break;
            case 'temperature_forecast':
                responses.push(temperatureForecastResponses[Math.floor(Math.random() * temperatureForecastResponses.length)]);
                break;
            default:
                responses.push("Desculpe, não entendi sua pergunta. Pode reformular?");
                break;
        }
    });

    return responses.join(" ");
}



export { intentClassifier, train, extractKeywords, getWeatherResponse };

File: clima/src/python/chatbot.py
import random
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.multiclass import OneVsRestClassifier
from sklearn.linear_model import SGDClassifier
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import MultiLabelBinarizer
import nltk
from nltk.tokenize import word_tokenize
import spacy

# Baixar recursos necessários do nltk
nltk.download('punkt')

# Carregar modelo de linguagem do spacy
nlp = spacy.load('pt_core_news_sm')

# Função para extrair palavras
def word_extractor(text):
    words = word_tokenize(text.lower())
    return words

# Treinamento do classificador
def train_classifier():
    train_data = [
        ("Como está o tempo hoje?", ["current_weather", "temperature_forecast"]),
        ("Qual é a previsão do tempo para amanhã?", ["weather_forecast"]),
        ("Está chovendo agora?", ["current_weather"]),
        ("Vai chover amanhã?", ["weather_forecast"]),
        ("Qual a temperatura atual?", ["current_temperature"]),
        ("Qual será a temperatura amanhã?", ["temperature_forecast"]),
        ("Faz calor amanhã?", ["temperature_forecast"]),
        ("Faz frio aqui", ["current_weather", "current_temperature"]),
        ("Vai nevar em São Paulo na quarta?", ["weather_forecast"]),
        ("Qual é a temperatura agora em Nova York?", ["current_temperature"]),
        ("Vai chover no Rio de Janeiro amanhã?", ["weather_forecast"]),
        ("Está nublado em São Paulo?", ["current_weather"]),
        ("Vai fazer frio na quarta-feira?", ["temperature_forecast"]),
        ("Previsão do tempo para hoje à noite", ["weather_forecast", "temperature_forecast"]),
        ("Quantos graus está fazendo agora?", ["current_temperature"]),
        ("Qual será a máxima de amanhã?", ["temperature_forecast"]),
        ("Quais são as chances de chuva hoje?", ["current_weather", "weather_forecast"]),
        ("Como estará o tempo no fim de semana?", ["weather_forecast", "temperature_forecast"]),
        ("Vai nevar no inverno?", ["weather_forecast", "temperature_forecast"]),
        ("Qual é a sensação térmica agora?", ["current_temperature"]),
        ("O que esperar do clima para a próxima semana?", ["weather_forecast", "temperature_forecast"]),
        ("Como estará o clima no Natal?", ["weather_forecast", "temperature_forecast"]),
        ("Vai ventar forte hoje?", ["current_weather"]),
        ("A umidade está alta agora?", ["current_weather"]),
        ("Quando será a próxima tempestade?", ["weather_forecast"]),
        ("Qual será a temperatura na semana que vem?", ["temperature_forecast"]),
        ("Vai chover na semana que vem?", ["weather_forecast"]),
        ("Como estava o tempo no mês passado?", ["current_weather"]),
        ("Qual era a temperatura no mês passado?", ["current_temperature"]),
        ("Qual é a previsão do tempo para a próxima semana?", ["weather_forecast"]),
        ("Vai nevar na próxima semana?", ["weather_forecast"]),
        ("Como estará o clima na próxima semana?", ["weather_forecast", "temperature_forecast"]),
        ("Estava muito quente no mês passado?", ["current_temperature"]),
        ("Qual será a máxima da próxima semana?", ["temperature_forecast"]),
        ("A previsão para a semana que vem é de chuva?", ["weather_forecast"]),
        ("A previsão para a próxima semana é de calor intenso?", ["weather_forecast", "temperature_forecast"]),
        ("Semana que vem vai fazer tanto calor quanto fez hoje?", ["weather_forecast", "temperature_forecast", "current_temperature"]),
        ("Amanhã qual será a temperatura em São Paulo?", ["temperature_forecast"]),
        ("Amanhã qual a temperatura na Baixada?", ["temperature_forecast"]),
        ("Amanhã qual a temperatura na Cidade?", ["temperature_forecast"]),
        ("Semana que vem qual a temperatura na Cidade?", ["temperature_forecast"]),
        ("Terça feira qual a temperatura no Galeão?", ["temperature_forecast"]),
        ("Segunda-feira qual a temperatura em Santos Lagos?", ["temperature_forecast"]),
        ("Na próxima segunda-feira vai chover no Rio de Janeiro?", ["weather_forecast"]),
        ("No próximo mês teremos dias quentes em Brasília?", ["weather_forecast", "temperature_forecast"]),
        ("Na semana que vem, como estará o clima em Salvador?", ["weather_forecast", "temperature_forecast"]),
        ("Qual é a previsão do tempo para o próximo fim de semana?", ["weather_forecast"]),
        ("Ontem, como estava o clima em Recife?", ["current_weather"]),
        ("Próxima quarta-feira terá temperaturas baixas em Florianópolis?", ["temperature_forecast"]),
        ("No próximo feriado, vai chover em Belo Horizonte?", ["weather_forecast"]),
        ("No próximo dia útil, qual será a temperatura em Manaus?", ["temperature_forecast"]),
        ("Daqui a dois dias, como estará o tempo em Porto Alegre?", ["weather_forecast", "temperature_forecast"])
    ]

    X_train = [data[0] for data in train_data]
    y_train = [data[1] for data in train_data]
    
    vectorizer = CountVectorizer(tokenizer=word_extractor, binary=True)
    X_train_counts = vectorizer.fit_transform(X_train)
    
    mlb = MultiLabelBinarizer()
    y_train_bin = mlb.fit_transform(y_train)
    
    classifier = OneVsRestClassifier(SGDClassifier(max_iter=1000, tol=1e-3))
    classifier.fit(X_train_counts, y_train_bin)
    
    return vectorizer, classifier, mlb

vectorizer, classifier, mlb = train_classifier()

# Função para extrair palavras-chave
def extract_keywords(question):
    doc = nlp(question.lower())
    date_match = [ent.text for ent in doc.ents if ent.label_ == "DATE"]
    location_match = [ent.text for ent in doc.ents if ent.label_ == "GPE"]
    weather_terms = ["clima", "tempo", "temperatura", "calor", "frio", "chuva", "chove", "chover", "sol", "nublado", "céu", "brisa", "neve", "neva", "nevar", "vento", "ventar", "umidade", "precipitação", "ventania", "graus", "pressão atmosférica", "tempestade", "nascer do sol", "pôr do sol"]
    has_weather_term = any(term in question.lower() for term in weather_terms)
    
    return {
        'date': date_match[0] if date_match else None,
        'location': location_match[0] if location_match else None,
        'hasWeatherTerm': has_weather_term
    }

# Função para gerar resposta baseada nas intenções classificadas e palavras-chave extraídas
def get_weather_response(intents, keywords):
    responses = []
    local = f" em {keywords['location']} " if keywords['location'] else ""
    
    for intent in intents:
        if intent == 'current_weather':
            responses.append(f"O tempo{local}está claro e ensolarado.")
        elif intent == 'weather_forecast':
            date = keywords['date'] if keywords['date'] else "amanhã"
            responses.append(f"A previsão para {date}{local}é de chuva com possibilidade de trovoadas.")
        elif intent == 'current_temperature':
            temperature = random.randint(23, 37)
            responses.append(f"A temperatura atual{local}é de {temperature}°C.")
        elif intent == 'temperature_forecast':
            date = keywords['date'] if keywords['date'] else "amanhã"
            temperature = random.randint(12, 37)
            responses.append(f"A temperatura {date}{local}será em torno de {temperature}°C.")
        else:
            responses.append("Desculpe, não entendi sua pergunta. Pode reformular?")
    
    return " ".join(responses)





question = "Qual é a previsão do tempo para amanhã?"
keywords = extract_keywords(question)
X_test_counts = vectorizer.transform([question])
predicted_intents_bin = classifier.predict(X_test_counts)
predicted_intents = mlb.inverse_transform(predicted_intents_bin)
response = get_weather_response(predicted_intents[0], keywords)
print(response)


question = "Chove hoje?"
keywords = extract_keywords(question)
X_test_counts = vectorizer.transform([question])
predicted_intents = classifier.predict(X_test_counts)
response = get_weather_response(predicted_intents[0], keywords)
print(response)

question = "Chove amanhã em juiz de fora?"
keywords = extract_keywords(question)
X_test_counts = vectorizer.transform([question])
predicted_intents = classifier.predict(X_test_counts)
response = get_weather_response(predicted_intents[0], keywords)
print(response)

